core/src/main/java/com/yahoo/oak/InternalOakBasics.java

```diff
@@ -171,6 +173,139 @@ protected UnscopedValueBufferSynced getValueUnscopedBuffer(ThreadContext ctx) {
         return new UnscopedValueBufferSynced(ctx.key, ctx.value, this);
     }
 
+    // Iterator State base class
+    static class BasicIteratorState<K, V> {
+
+        private BasicChunk<K, V> chunk;
+        private BasicChunk.BasicChunkIter chunkIter;
+        private int index;
+
+        public void set(BasicChunk<K, V> chunk, BasicChunk.BasicChunkIter chunkIter, int index) {
+            this.chunk = chunk;
+            this.chunkIter = chunkIter;
+            this.index = index;
+        }
+
+        protected BasicIteratorState(BasicChunk<K, V> nextChunk, BasicChunk.BasicChunkIter nextChunkIter,
+                                     int nextIndex) {
+
+            this.chunk = nextChunk;
+            this.chunkIter = nextChunkIter;
+            this.index = nextIndex;
+        }
+
+        BasicChunk<K, V> getChunk() {
+            return chunk;
+        }
+
+        BasicChunk.BasicChunkIter getChunkIter() {
+            return chunkIter;
+        }
+
+        public int getIndex() {
+            return index;
+        }
+
+    }
+
+
+    /************************/
+    /* Basic Iterator class */
+    /************************/
+    abstract class BasicIter<T> implements Iterator<T> {
+
+
+        /**
+         * the next node to return from next();
+         */
+        private BasicIteratorState<K, V> state;
+
+        /**
+         * An iterator cannot be accesses concurrently by multiple threads.
+         * Thus, it is safe to have its own thread context.
+         */
+        protected ThreadContext ctx;
+
+        /**
+         * Initializes ascending iterator for entire range.
+         */
+        BasicIter() {
+            this.ctx = new ThreadContext(keysMemoryManager, valuesMemoryManager);
+        }
+
+        public boolean hasNext() {
+            return (state != null);
+        }
+
+        protected abstract void initAfterRebalance();
+
+
+        // the actual next()
+        public abstract T next();
+
+        /**
+         * Advances next to higher entry.
+         *  previous index
+         *
+         * The first long is the key's reference, the integer is the value's version and the second long is
+         * the value's reference. If {@code needsValue == false}, then the value of the map entry is {@code null}.
+         */
+        void advance(boolean needsValue) {
+            boolean validState = false;
+
+            while (!validState) {
+                if (state == null) {
+                    throw new NoSuchElementException();
+                }
+
+                final BasicChunk<K, V> chunk = state.getChunk();
+                if (chunk.state() == BasicChunk.State.RELEASED) {
```

Comment:
Please add a TODO comment not to access the keys on the RELEASED chunk once the key might be released